Remove `transport_versions` from cluster state API

DaveCTurner opened the pull request:
Since #99114 the cluster state API exposes per-node versions such as
their transport versions under `node_versions` so there's no need to
duplicate this information under `transport_versions`. This commit
removes the unnecessary information.

elasticsearchmachine commented:
Hi @DaveCTurner, I've created a changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

DaveCTurner committed:
Remove `transport_versions` from cluster state API

Since #99114 the cluster state API exposes per-node versions such as
their transport versions under `node_versions` so there's no need to
duplicate this information under `transport_versions`. This commit
removes the unnecessary information.

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

thecoop commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java:
```diff
@@ -12,70 +12,135 @@
+            case UPGRADED -> {
+                assertFalse(description, hasTransportVersions);
+                assertTrue(description, hasNodesVersions);
+                assertTrue(description, versionsByNodeId.values().stream().allMatch(v -> v.equals(Version.CURRENT)));
```
`everyItem` matcher is a more declarative way of doing this

DaveCTurner replied:
++ TIL

thecoop commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java:
```diff
@@ -12,70 +12,135 @@
-        Map<String, Object> masterResponse = entityAsMap(client().performRequest(request));
-        assumeFalse("Master needs to not know about transport versions", masterResponse.containsKey("transport_versions"));
+        // waitUntil because the versions fixup on upgrade happens in the background so may need a retry
+        assertTrue(waitUntil(() -> {
```
Will the default timeout be enough here on slow test clusters?

DaveCTurner replied:
In current `main` we don't retry at all and I couldn't find any failures in CI history so I think this will be fine.

thecoop commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java:
```diff
@@ -12,70 +12,135 @@
-        Map<?, TransportVersion> tvs = ((List<?>) response.get("transport_versions")).stream()
-            .map(o -> (Map<?, ?>) o)
-            .collect(Collectors.toMap(m -> m.get("node_id"), m -> TransportVersion.fromString(m.get("transport_version").toString())));
+            assertEquals(description, nodeIds.size(), clusterState.evaluateArraySize("transport_versions"));
```
Can use `hasSize` matcher here

DaveCTurner replied:
Not without doing some unchecked casting or reversing the expected/actual values I think.

thecoop commented:
This changes the test purpose. The test was specifically to check that when upgrading from <8.8.0 to >=8.8.0, AND the master was <8.8.0, the min transport version was inferred, then it is replaced with the real one.

Having this more general test is really good, but I would prefer to keep an explicit test for the transport version fixup logic

DaveCTurner committed:
Merge branch 'main' into 2023/09/06/remove-cluster-state-transport-versions

DaveCTurner committed:
Comparable

DaveCTurner committed:
everyItem

DaveCTurner commented on x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java:
```diff
@@ -12,70 +12,135 @@
-        assumeTrue("Only runs on the upgraded cluster", CLUSTER_TYPE == ClusterType.UPGRADED);
-        // once everything is upgraded, the master should fill in the real transport versions
+        if (hasTransportVersions) {
+            assertTrue(description, UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS));
```
Ah didn't realise it was `Comparable`, nice.

DaveCTurner commented:
I'm not convinced the previous test really did test that: it seemed very likely it would be skipped without really asserting anything, and it wasn't really doing anything that looked at whether the master was upgraded or not.

Or, put differently, I'm not sure I see how this new test could pass without doing the correct thing re. inferred versions. It verifies that we always see either all the versions or none of them, and if the coordinating node is upgraded then it will always emit a `nodes_versions` array so we must see all of them, and therefore some of them must be inferred since it got its result from the master which may or may not be upgraded.

thecoop commented:
I suppose - in that case it would be good to have some comments at various points on what behaviour the checks are specifically for - so that if they fail, the dev is pointed to the fixup classes.

DaveCTurner committed:
Commentary

DaveCTurner commented:
++ makes sense, I added some commentary in 0083b8c.

thecoop reviewed (approved):
Looks great, very thorough
